Bug 708984 - Promoting SDK promise library for toolkit.

Gozala opened the pull request:
Please don't close this pull request as it's only for review purposes. There is no intention of landing this change.

Gozala committed:
Promoting SDK promise library for toolkit.

Gozala commented:
> Your solution is quite elegant in terms of message-passing/functional programming.

Thank you

> However, in the context, I cannot find any good reason to not store the result directly but rather prefer a pseudo-
> promise/agent that can pass it as message to further observers. 

Promises are all about message passing :)

> This would make perfect sense in Erlang or if the result > were somehow stored in another process/node, but here, the
> main consequence is that it complicates considerably the execution path and it makes it much harder for most 
> newcomers/potential contributors to understand what is going on.

I'm not convinced that this opinion is true, in fact implementation is way simpler and there is much smaller surface
for bugs, just remember how many iterations it took you to get all the semantics right (still have to verify that it's completely right now). This implementation avoids dreaded logic and there for is ways simpler. I do understand you don't agree, but
that is why I asked others to provide a feedback to a get a more objective opinion. Otherwise it feels to me that perfectly
well working implementation that had multiple reviews already was forked and made less maintainable (at least for me)
for no obvious reasons.

Gozala commented:
@Yoric It would be handy if you could put your comments in the pull request itself (in the diff tab) otherwise it's little hard to follow up.

Gozala committed:
Fix handler names.

Gozala committed:
Address @Yoric's review comments.

Gozala commented:
@Yoric could you please comment here https://github.com/mozilla/mozilla-central/pull/4/files

Yoric commented:
So, inch by inch, this is getting closer to the quality required for mozilla-central. Even if we omit our core disagreement on how to implement `result`, there is still work to do. Unfortunately, by now, I am just too tired to continue pulling you kicking and screaming to that level of quality.

Gozala commented:
> So, inch by inch, this is getting closer to the quality required for mozilla-central. Even if we omit our core disagreement on how to > implement result, there is still work to do. Unfortunately, by now, I am just too tired to continue pulling you kicking and screaming > to that level of quality.

I think your comment above is really inappropriate.

Yoric commented:
It could be. If so, please accept my apologies.

I would like you to realize, however, that your code (this includes comments and variable naming) really does not meet the quality level for inclusion mozilla-central. For the past three months or so, I have tried to get you to make the necessary changes to reach production quality. So far, almost every single change I have asked has been either ignored or nitpicked to death. I have chosen to disregard the tone of a number of your replies, which was often borderline insulting, and assume that this was a linguistic issue.

While your remarks have been right in several occasions, this process has been made needlessly tiresome.

Hence, as mentioned above, I am just too tired and I am planning to stop reviewing/improving your code.

Gozala commented:
> It could be. If so, please accept my apologies.

Accepted

> I would like you to realize, however, that your code (this includes comments and variable naming) really does not meet
> the quality level for inclusion mozilla-central. For the past three months or so, I have tried to get you to make the
> necessary changes to reach production quality. So far, almost every single change I have asked has been either ignored or
> nitpicked to death.

I disagreed as code passed standard mozilla review process ensuring its quality. Still, I have agreed to make
necessary naming & comment changes, although I mentioned that I was occupied with more important tasks
concerning Q2 goals of the project. Also I suggested you to make appropriate changes it could not wait. 

> I have chosen to disregard the tone of a number of your replies, which was often borderline insulting, and assume that this
> was a linguistic issue.

It was never my intention to make you feel that way, if so my apologies. Please indicate if you ever notice that again, I'd like to make sure I phrase my thoughts properly. 

> While your remarks have been right in several occasions, this process has been made needlessly tiresome.

I have exact same experience on my side. From the very beginning I mentioned that fork was unnecessary and all
issues could have being addressed without: https://github.com/mozilla/addon-sdk/pull/433#issuecomment-5604214
Still you were convinced that implementation could be simplified, so I followed along to illustrate all the details that you
were not taken into account. As I expected, after addressing those details implementation ended up even more complicated.

Yoric commented:
> I disagreed as code passed standard mozilla review process ensuring its quality

I would like to clarify one thing: barring any mistake, I never wrote that your code was not up to _mozilla_ quality, but that it was not up to _mozilla-central_ quality. I now realize that it may be ambiguous and/or that you may be unaware of this, but distinct subprojects of Mozilla have sometimes very different requirements and notions of quality.

> I have exact same experience on my side. From the very beginning I mentioned that fork was unnecessary and all
> issues could have being addressed without: https://github.com/mozilla/addon-sdk/pull/433#issuecomment-5604214

Just to put things in context, you may recall that several months before you landed your implementation of promises, I attempted to get in touch with you, exactly to avoid the current situation, as there were already two implementations of promises in the Mozilla codebase. At the exact same time, I got in touch with Joe Walker and we successfully merged our implementations. I only heard back from you once we had reached the current problem, with two implementations of promises, two codebases depending on them, and two different sets of semantics.

You may also recall that a few weeks ago, I have submitted for your review a version of the code that did not touch your semantics and that simply clarified your implementation by touching only documentation and variable/arguments naming. This was meant as a first step towards integration of your code in mozilla-central. To this day, you still have not reviewed this commit.

Until yesterday, your message was, rather clearly, that you rejected any form of change, often quite violently. For some proposed changes, you were right, for others not. If this is not what you meant, then we obviously have (yet another) communication issue.

P.S.: Apologies accepted, too, and thank you.

Yoric commented:
> @Yoric could you please comment here https://github.com/mozilla/mozilla-central/pull/4/files

Sorry about that.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+/*jshint undef: true es5: true node: true browser: true devel: true
+         forin: true latedef: false */
+/*global define: true, Cu: true, __URI__: true */
+;(function(id, factory) { // Module boilerplate :(
```
What's that `;`?

Gozala replied:
Some minifiers damage code when contactinating multiple files. If one this is appended to has no `;` in the end such minifiers break code. Although I understand it's not very relevant in this context and I can remove it if you want.

Yoric replied:
I don't particularly mind that one, I was just wondering if it was a typo.

Yoric commented:
Since apologies have been exchanged, I will give it another try.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+    define(factory);
+  } else if (typeof(require) === 'function') { // CommonJS
+    factory.call(this, require, exports, module);
+  } else if (~String(this).indexOf('BackstagePass')) { // JSM
```
Really, please get rid of this `~` shortcut. It hinders readability.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+
+'use strict';
+
+function resolution(value) {
```
Comments need rephrasing.

Try something along the lines of

```
// Internal utility: Wrap `value` as a simplified promise that always resolves to `value`.
// Note the result is not a full promise, as its method `then` does not return anything.
```

Yoric replied:
Also, I am not convinced by name `resolution`. Maybe `asResolvedPromise`?

Gozala replied:
I don't really like `asResolvedPromise` and `resolution` seems natural to me. If you don't mind I'll leave it for now and let native English @Mossop  decide weather naming makes sense and if it isn't hopefully get some suggestions.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+      this['Components'].utils.import(uri, imports);
+      return imports;
+    }, this, { uri: __URI__, id: id });
+    this.EXPORTED_SYMBOLS = Object.keys(this);
```
I have no clue what this `uri` is for.

Yoric replied:
Also, a .jsm will generally export only one symbol which will act as a namespace. Perhaps `Promise`?

Gozala replied:
> I have no clue what this uri is for.

`uri` just provides `module.uri` expected by commonjs modules.
I do agree this  does not necessary makes much sense in this particular case, but since some SDK modules are wrap in such module boilerplate I'd prefer to keep that boilerplate consistent  across these modules. So unless you feel very strongly about this I'd really prefer to keep it.

Gozala replied:
> Also, a .jsm will generally export only one symbol which will act as a namespace. Perhaps Promise?

Quick MXR search indicated bunch of modules that don't follow this pattern:
https://mxr.mozilla.org/mozilla-central/search?string=EXPORTED_SYMBOLS

Also this maps much better to the way CommonJS modules work, so I'd rather leave it as is. Specially since it's always possible to import JSM into `Promise` if desired so.

Yoric replied:
As you wish.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+  return { then: function then(resolve) { resolve(value); } };
+}
+
+function rejection(reason) {
```
Same remarks as `resolution`.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+  return { then: function then(resolve, reject) { reject(reason); } };
+}
+
+function attempt(f) {
```
Since you use DSL to describe some of my functions, I feel compelled to remark that `attempt` does not look any better.
It takes a function and returns a function that sometimes returns a value and sometimes a promise. This is certainly irregular.

Yoric replied:
Also, `options` is an ill-chosen name as the argument is not an option. Perhaps `args`?

Yoric replied:
Nit: Also, m-c coding guidelines tend to insist that formatting should look like

```
try { // newline here
   ...
} catch (error) {  // whitespace after `catch`
  ...
}
```

Gozala replied:
I'll rename `args` to `input` hope that's fine with you. I don't like `args` as it's just one.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+  - `reject` Single shot function that rejects returned `promise` with a given
+    `reason` argument.
+
+  ## Examples
```
If you do not want to improve the example, just remove it.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+  deferred.resolve(value)
+  */
+
+  // Create an array where `pending` observers will be collected. Once promise
```
Split these comments in two.

Something along the lines of:

```
// An array of observers of this promise. May be either
// - an array of pairs of observers (if the promise is not fulfilled yet);
// - or `null` (once the promise is fulfilled).
var pending = [];
```

Yoric replied:
```
// The resolution value for this promise. May be either
// - `undefined` (if the promise is not fulfilled yet)
// - a promise `v` (if the promise was resolved with a promise `v`)
// - a pseudo-promise (if the promise was rejected or resolved with something other than a promise)
var result;
```

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+        // This will make sure that `deferred` is rejected on exceptions in
+        // event handlers. If handler is not provided fall back to
+        // `resolution` / `rejection` that will just propagate values.
+        onResolve = resolve ? attempt(onResolve) : resolution;
```
That line looks fishy. Did you mean `onResolve = onResolve ? ... : ...` and `onReject = onReject ? ... : ...`?

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+        // event handlers. If handler is not provided fall back to
+        // `resolution` / `rejection` that will just propagate values.
+        onResolve = resolve ? attempt(onResolve) : resolution;
+        onReject = reject ? attempt(onReject) : rejection;
```
I insist that the new `onResolve`/`onReject` are _not_ normalized forms of arguments `onResolve`/`onReject`. Using the same name for two functions that have different signatures encourages misreadings.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+        onResolve = resolve ? attempt(onResolve) : resolution;
+        onReject = reject ? attempt(onReject) : rejection;
+
+        // Create an observers that will invoke registered handlers
```
For the fourth time: either _an observer_ (singular) or _a pair of observers_ (plural)!

Gozala replied:
Please bare with me, it's little hard to aggregate comments from multiple different sources (threads you commented in).

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+
+        // If promise is pending register listeners. Otherwise forward
+        // observers to a result.
+        if (pending) pending.push([ resolve, reject ]);
```
Nit: on mozilla-central, we generally write

```
if (pending) {
  ...
} else {
  ...
}
```

This makes it easier to patch the code or add debugging/logging support without introducing errors.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+      Resolves associated `promise` to a given `value`, unless it's already
+      resolved or rejected.
+      **/
+      if (pending) {
```
Nit: in mozilla-central, we generally write

```
if (!pending) {
  return;
}
```

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+        // Store resolution `value` as a promise (`value` itself may be a
+        // promise), so that all subsequent listeners can be forwarded to it,
+        // which either resolves immediately or forwards if `value` is
+        // a promise. This takes care of value / error propagation.
```
Comments need some rephrasing. Maybe something along the lines of:

```
// Store resolution `value` in `result` as a promise, so that all subsequent listeners can be
// forwarded to it, whether `value` or not is already a promise.
```

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+        // Mark promise as resolved.
+        pending = null;
+        // Forward resolution value to all registered observers.
+        while (index < count) result.then.apply(result, observers[index++]);
```
Nit: As above, in m-c, we would rather write

```
while (index < count) {
  // ...
}
```

Yoric replied:
Also, this is a `while` loop with the following characteristics:
- create and initialize a variable;
- continue while this variable fulfills some condition;
- at the end of the loop, increment the observer.

Generally, such loops are rather written as `for` loops.

Among other things, it makes easier to figure out the value of `index++`.

Gozala replied:
In the end I had to switch to the `while` with `.shift()` I indicate reasons in code comments and in other comment here.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+        // a promise. This takes care of value / error propagation.
+        result = isPromise(value) ? value : resolution(value);
+        var observers = pending, index = 0, count = observers.length;
+        // Mark promise as resolved.
```
I would add "_to ensure that observers to not register additional observers_" or something such.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+      Rejects associated `promise` with a given `reason`, unless it's already
+      resolved / rejected.
+      **/
+      deferred.resolve(rejection(reason));
```
```
// Convert `reason` to a rejected promise and use the general mechanisms of `resolve`.
```

Yoric replied:
Also, I suspect that readers will find it a little weird that `reject` only calls `resolve`. I certainly did.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+    var globals = this
+    factory(function require(id) {
+      return globals[id];
+    }, (globals[id] = {}), { uri: document.location.href + '#' + id, id: id });
```
Again, no clue what this `uri` is for.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,162 @@
+      var imports = {};
+      this['Components'].utils.import(uri, imports);
+      return imports;
+    }, this, { uri: __URI__, id: id });
```
I don't think that `__URI__` is even documented anywhere. Are you sure it is a good idea to use it?

Yoric commented:
So, as mentioned above, this is getting better, inch by inch.

Unsurprisingly, the logic of your implementation is no simpler than that of mine, but at least, with a smaller package (something you have done already), better documentation and naming (something we are working on), users have a fighting chance of understanding the code.

Gozala committed:
Address more review comments.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+ *
+ *  ## Example
+ *
+ *  function readURI(uri, type) {
```
This example is much better. I would just rename it `fetchURI` or `getURI` instead of `readURI`.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+
+/**
+ * Internal utility: Wraps given `value` into simplified promise, pre-resolved
+ * to `value`. Note the result is not a full promise, as it's method `then`
```
Grammar: not "it's", "its".

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+
+/**
+ * Internal utility: Wraps given input into simplified promise, pre-rejected
+ * with given `reason`. Note the result is not a full promise, as it's method
```
As above.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+'use strict';
+
+/**
+ * Internal utility: Wraps given `value` into simplified promise, pre-resolved
```
"into _a_ simplified promise"

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+/**
+ * Internal utility: Wraps given `value` into simplified promise, pre-resolved
+ * to `value`. Note the result is not a full promise, as it's method `then`
+ * does not returns anything.
```
Grammar: not "returns", "return".

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+}
+
+/**
+ * Internal utility: Decorates given `f` function, so that on exception promise
```
That sentence doesn't mean anything.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+  return function effort(input) {
+    try {
+      return f(input);
+    }
```
Nit: This is still not the layout recommended on m-c.

Gozala replied:
I'm little tired of this styling issues. In SDK we explicitly forbid cuddleing both `catch` and `else`. As a matter of fact we also allow single line `if else` without braces and many more. I'm just pointing this out here because SDK is on it's way to mozilla-central and there were no plan for additional reviews or stylistic changes.

Now that being said, I really can't care less about anything, but style. It's just switching back and forth between incompatible styles is painful and there is no way to share code that would be compliant with both styles.

Please lets defer all the styling issues for now and if SDK code has to be trans-styled we'll do it all together.

Yoric replied:
ok

Mossop replied:
The style guide is exactly that, a guide. As long as new code is along the right lines it is a waste of time to spend too much time over nit-picking it. The important things are that the code is consistent with the module it lives in and that it is readable. This particular style (not cuddling else and catch) is used quite a bit in Mozilla code anyway and is just fine.

Yoric replied:
As mentioned above: "ok".

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+
+/**
+ * Internal utility: Returns `true` if given `value` is promise. Value is
+ * assumed to be promise if it implements `then` method.
```
"if it implements method `then`."

Yoric replied:
"to be _a_ promise"

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+}
+
+/**
+ * Internal utility: Returns `true` if given `value` is promise. Value is
```
"if `value` is a promise"

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+
+/**
+ * Creates deferred object containing fresh promise & methods to either resolve
+ * or reject it. Result contains following properties:
```
"_The_ result is an object with the following properties:"

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+ *  }
+ */
+function defer() {
+  // Array to store promise observers up until it's resolved.
```
"until _the promise_ is resolved"

Yoric replied:
Also, the contents of the array are not very clear.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+function defer() {
+  // Array to store promise observers up until it's resolved.
+  var observers = [];
+  // Promise `result`, which will be assigned a resolution value once resolved.
```
This is (grammatically) ambiguous.

"Once `deferred.promise` is resolved, `result` is assigned a resolution value."

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+  // Array to store promise observers up until it's resolved.
+  var observers = [];
+  // Promise `result`, which will be assigned a resolution value once resolved.
+  // Note that result will always be assigned promise (or alike) object to take
```
"or alike" is an anti-explanation. It just means "find out for yourself", which is not the objective of a comment.

Yoric replied:
I assume you meant `result` and not just "result".

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+  var observers = [];
+  // Promise `result`, which will be assigned a resolution value once resolved.
+  // Note that result will always be assigned promise (or alike) object to take
+  // care propagation through promise chains. If `result` is `null` then
```
Do not expect people to know what a "promise chain" is at this stage.

Yoric replied:
"care _of_ propagation" - although I suspect that the whole sentence could be improved

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+      then: function then(onResolve, onReject) {
+        var deferred = defer();
+        // Decorate `onResolve` / `onReject` handlers with `attempt` wrapper.
+        // This way it's guaranteed to always returns even on exceptions. On
```
Pronoun "it" is ambiguous in this context. I assume you mean "These decorated functions always return, even in case of exception" or something along these lines.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+        onResolve = onResolve ? attempt(onResolve) : resolution;
+        onReject = onReject ? attempt(onReject) : rejection;
+
+        // Create a pair of observers that invoke given handlers & propagate
```
Victory, you have finally fixed that grammar error :)

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+        var deferred = defer();
+        // Decorate `onResolve` / `onReject` handlers with `attempt` wrapper.
+        // This way it's guaranteed to always returns even on exceptions. On
+        // exceptions promise rejected with a thrown error is returned.
```
"In case of exception, instead of throwing, the decorated functions return a rejected promise holding the exception."

Yoric replied:
However, expect that readers will have no clue why in case of success, the decorated functions do not return a promise.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+        onReject = onReject ? attempt(onReject) : rejection;
+
+        // Create a pair of observers that invoke given handlers & propagate
+        // results to a resulting promise (One that is returned by `then`).
```
Reading this sentence, I have no clue which promise you mean.

Gozala replied:
Do you have suggestion how to rephrase this ? I'm sorry it's just no matter how much I try to make best comments I'm still non native English speaker.

Yoric replied:
Maybe "Create a pair of observers that invoke `onResolve`/`onReject` and propagate the results to `deferred.promise`."

Yoric replied:
However, you have two `deferred` (hence two `deferred.promise`) in the scope (one masking the other), which makes it ambiguous. You may want to rename the inner `deferred`. Right now, I have no clear-cut idea regarding how you could call it, perhaps `deferChain` or `nextDefer` or something such.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+        function resolve(value) { deferred.resolve(onResolve(value)); }
+        function reject(reason) { deferred.resolve(onReject(reason)); }
+
+        // If enclosed promise is resolved, then forward observers to the
```
If "enclosed promise" is `deferred.promise`, just name it, this will be less ambiguous.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+        function reject(reason) { deferred.resolve(onReject(reason)); }
+
+        // If enclosed promise is resolved, then forward observers to the
+        // resolution result (which is promise or alike).
```
As above: "or alike" is a slap in the face of readers.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+
+        // If enclosed promise is resolved, then forward observers to the
+        // resolution result (which is promise or alike).
+        if (result)
```
I do not know if you want to to look like mozilla-central code, but if you do, the curly brackets are recommended, for the reasons mentioned previously.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+          result.then(resolve, reject);
+        // Otherwise register observers.
+        else
+          observers.push({ resolve: resolve, reject: reject });
```
Using an object instead of an array is, indeed, nicer. I assumed you used an array for performance reasons, didn't you?

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+    resolve: function resolve(value) {
+      if (!result) {
+        // Store resolution `value` as a `result` of enclosed promise. Note
+        // that `result` is a promise (like) or normalized to one, this way all
```
Third slap.

Yoric replied:
There is nothing simple about it, don't pretend there is :)

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+     */
+    resolve: function resolve(value) {
+      if (!result) {
+        // Store resolution `value` as a `result` of enclosed promise. Note
```
This sentence does not mean anything.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,185 @@
+     * resolved or rejected.
+     */
+    resolve: function resolve(value) {
+      if (!result) {
```
So, I take it you do not like the recommended mozilla-central style I suggested in one of the previous versions of this pull request?

Gozala committed:
Switch back to observer shifting otherwise FIFO propagation order can't be guaranteed.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,192 @@
+     */
+    resolve: function resolve(value) {
+      if (!result) {
+        // Store resolution `value` in a `result` as a promise, so that all
```
I strongly suspect that readers of the code don't want to know that it is "automatic". They want to know how it works. If you do not want the comments I suggested in your previous pull request, please find something else that actually explains how it works, rather than telling readers that they should not care.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,192 @@
+      if (!result) {
+        // Store resolution `value` in a `result` as a promise, so that all
+        // the subsequent handlers can be simple forwarded to it. This way all
+        // the promise propagation will be automatically taken care of.
```
I suspect you mean the opposite: "Forward a result to all registered observers".

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,192 @@
+        // Store resolution `value` in a `result` as a promise, so that all
+        // the subsequent handlers can be simple forwarded to it. This way all
+        // the promise propagation will be automatically taken care of.
+        result = isPromise(value) ? value : resolution(value);
```
Oh, you have changed the algorithm to copy the array. Why is that?

Yoric replied:
Also, you now never nullify `pending`. This is a bad practice, memory-wise.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,192 @@
+        // the subsequent handlers can be simple forwarded to it. This way all
+        // the promise propagation will be automatically taken care of.
+        result = isPromise(value) ? value : resolution(value);
+        // Forward already registered observers to a `result` promise in order
```
Ok, ok, I give up on trying to make you write this `for` loop with a `for` :)

Gozala commented:
@Yoric Please note that I did not mean to go back to square one with last change. It's just I got mislead by our previous conversations and was convinced that simple loop would be ok. Later I recalled why I used `.shift()` instead, and hopefully documented that properly this time. I also believe your fork wont have FIFO prioritization for the same reasons.

Gozala commented:
@Yoric Also please be little patient :) If I don't address all comments in the same commit that does not means I'm not going to. I'll post back once I consider that patch is ready for the next round. I will also clearly indicate by comments if I don't agree on a particular note.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,192 @@
+        result = isPromise(value) ? value : resolution(value);
+        // Forward already registered observers to a `result` promise in order
+        // of their registration. Note we internally shift observer pairs from
+        // queue until it's exhausted to guarantee
```
Grammar: Not "it's", "its".

Gozala committed:
Make difference between promise resolution and fulfillment more obvious.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,192 @@
+        // Forward already registered observers to a `result` promise in order
+        // of their registration. Note we internally shift observer pairs from
+        // queue until it's exhausted to guarantee
+        // [FIFO](http://en.wikipedia.org/wiki/FIFO) prioritization.
```
I see no reason to guarantee any order of invocation of listeners. Unless it is a requirement of some set of specifications, this looks a lot like overkill and/or encouraging an anti-pattern.

Yoric replied:
On the other hand, I am pretty sure that a reader who ventures in the source code of promises knows what a FIFO is (or at least knows enough to look for the definition). I don't think that the link to Wikipedia adds anything.

Gozala replied:
That is how **Q** behaves and in fact it's pretty important let me give you an example why:

lets say we you have two promises `a` and `b`. Two handlers `h1` and `h2` are registered on a promise `a`.
Handler `h1` has side effect of registering `h3` handler on the promise `a`, that attempts to resolve promise `b` with `foo`
Also `h2` handler attempts to resolve `b` but with `bar` instead.

One would expect that `b` is resolved with `bar` as `h2` was registered before `h3`. But if you do not guarantee FIFO order
`b` will be resolved with `foo` instead.

In real life cases are obviously way more complicated than that and reasoning about them becomes way harder if such order is not
guaranteed.

Not to even mention that promise resolution by itself is designed around "first come wins" bases so sticking to that design is highly
important.

Gozala replied:
> On the other hand, I am pretty sure that a reader who ventures in the source code of promises knows what a FIFO is (or at least knows enough to look > for the definition). I don't think that the link to Wikipedia adds anything.

I don't know you was expecting newcomers to JS contributing, so.... :)

Yoric replied:
_Insert sound of concurrent spider-sense tingling._

Tell me if I am wrong, but it is my understanding that one of the main reasons to introduce promises in Jetpack is to encourage/enforce asynchronous programming. If you start building code that relies on order of evaluation, you are effectively introducing hypotheses that will all fall down and crumble once your code is effectively asynchronous/multi-threaded/multi-process.

I strongly suggest avoiding these considerations like a plague.

Yoric replied:
> I don't know you was expecting newcomers to JS contributing, so.... :)

Sorry, I don't understand that remark.

Gozala replied:
Please ignore it.

Gozala replied:
> Tell me if I am wrong, but it is my understanding that one of the main reasons to introduce promises in Jetpack is to encourage/enforce asynchronous > programming. If you start building code that relies on order of evaluation, you are effectively introducing hypotheses that will all fall down and crumble > once your code is effectively asynchronous/multi-threaded/multi-process.

Let me try to put it other way, if you deal with async computations using promises order of handler calls will be FIFO,
now if you don't guarantee the same order in a non-async computations it easy to catch someone by a surprise.

Now even if you disagree with above statement I think it's important to be complaint with behavior of Q which happens to be FIFO.

Yoric replied:
Well, you claim that you intend to write client code that strongly depends on the order of evaluation of asynchronous code.

This has the following drawbacks:
- most contributors or users will not be aware of this guarantee, so they will not be able to understand what makes your code tick;
- in many cases, most contributors or users will have no clue of the order in which handlers have been registered, so this will again get in their path if they try and understand what makes your code tick;
- as a former concurrency guy myself, I tell you that this hypothesis is very shaky once you introduce actual concurrency, which I understand is one of the objectives of Jetpack.

> Now even if you disagree with above statement I think it's important to be complaint with behavior of Q which happens to be FIFO.

I have not been able to find this in the specifications of Q. Could you point me to it? If this is just an implementation detail and not a specification, I really suggest we avoid these semantics.

Gozala replied:
> This has the following drawbacks:
> 
>   most contributors or users will not be aware of this guarantee, so they will not be able to understand what makes your code
> tick; 

In most the case users don't even need to care about this, as long as they can move from sync to async without issues that's all they need to know really. As of contributors I think I tried to document fact that it guarantee FIFO order pretty thoroughly, if you think of any other ways to make contributors aware, I'm open to that. 

>  in many cases, most contributors or users will have no clue of the order in which handlers have been registered, so this will
> again get in their path if they try and understand what makes your code tick;

If your code does not cares about an order, that's great and there is no problem in knowing which order it works. But if there is a cases where your code depend on order it's really useful to have it deterministic. If someones code depends on such order I expect them to try to learn about guarantees that library they use makes.

> as a former concurrency guy myself, I tell you that this hypothesis is very shaky once you introduce actual concurrency, which I > understand is one of the objectives of Jetpack.

In Jetpack promises just help us to compute and reason about eventual values & computations. Having undetermined handler invoke order will make this harder. That way one can't reason about behavior without finding out if promise is resolved in the same turn of event loop or not.

> I have not been able to find this in the specifications of Q. Could you point me to it? If this is just an implementation detail and
> not a specification, I really suggest we avoid these semantics.

Q has it, as they always queue all the operations so it's never sync, that is why order there is FIFO.

Gozala replied:
Also you completely ignore some other arguments I made: 
- Both in JS and DOM order in which handlers are invoked is deterministic. Making API that invokes handlers in undeterministic order is pretty risky design choice that I expect to be harmful. 
- I already described a use case where undetermined order will force users into handling queuing manually. 

I still don't see strong arguments against making order deterministic. All I hear is:

> - users / contributors won't be aware

  I hardly see it as a problem, not all drivers know how their car engine works, but they still can use it to drive around.

> - That is a shaky bad idea.

  Sorry, but I'm not convinced, for reasons outlined before.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+ *  }
+ */
+function defer() {
+  // Define FIFO queue of observer pairs. Once promise is resolved & all queued
```
Well, it seems that they are not pairs anymore.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+        // `result` will be a promise all the value / error propagation will
+        // be automatically taken care of.
+        result = isPromise(value) ? value : fulfillment(value);
+        // Forward already registered observers to a `result` promise in the
```
Insert here usual bracket-related m-c nit.

Gozala replied:
> Insert here usual bracket-related m-c nit.

https://mxr.mozilla.org/mozilla-central/search?string=[^\%3F]%2B\%3F[^\%3A]%2B\%3A&regexp=1&find=.jsm&findi=&filter=^[^\0]*%24&hitlimit=&tree=mozilla-central

Yoric replied:
Search on MDN for _Mozilla coding guidelines_ :)

Gozala replied:
Quote from [Mozilla coding guidelines](https://developer.mozilla.org/En/Developer_Guide/Coding_Style)

> Of course every rule has an exception, but it's important to know the rules nonetheless!

I believe this case and others from mxr are well justified exceptions ;)

Yoric replied:
As you wish. As mentioned, this was minor stuff.

**edit** Fixed spelling mistake.

Mossop replied:
The only styling nit I'd make here is that a few newlines before each comment block would make this much more readable

Yoric replied:
Agreed.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+        // be automatically taken care of.
+        result = isPromise(value) ? value : fulfillment(value);
+        // Forward already registered observers to a `result` promise in the
+        // order they were registered. Note that we internally shift observer
```
Insert here months-old criticism about traversing an array in quadratic time.

Yoric replied:
Also, your semantics depends on the order of evaluation of arguments. This is seldom a good sign and this is always disconcerting for readers.

Gozala replied:
> Also, your semantics depends on the order of evaluation of arguments. This is seldom a good sign and this is always disconcerting for readers.

I'm ok with saving observes in `resolve, reject` pairs if you think that improves it, I have not viewed this as an issue, so was just avoiding more object allocations and property lookups.

Yoric replied:
Well, until yesterday, this is what you were doing. Then you starting putting them in objects, which is even nicer, albeit a little slower. And all of a sudden, you start optimizing for speed while we are attempting to optimize readability, which sounds like a bad idea.

Gozala replied:
I started joining them to an object just to avoid naming them with `resolve2, reject2`, as all other names were already taken :)

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+        function onReject(reason) { deferred.resolve(reject(reason)); }
+
+        // If enclosed promise observers queue is still alive a enqueue a new
+        // pair into it. Note that this does not necessary means that promise
```
Nit: Not a pair anymore. Now it is an irregular data structure that alternates stuff. I assume this is for performance reasons, and if so, this needs to be documented.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+        // pair into it. Note that this does not necessary means that promise
+        // is pending, it may already be resolved, but we still have to queue
+        // observers to guarantee an order of propagation.
+        if (observers) {
```
Nit: Usual bracket remark. Please tell me if you have no intention of doing it, I will stop with these nitpicks.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+        // the subsequent handlers can be simply forwarded to it. Since
+        // `result` will be a promise all the value / error propagation will
+        // be automatically taken care of.
+        result = isPromise(value) ? value : fulfillment(value);
```
Instead of this weird shift-based loop, if for some mysterious reason, you really want to guarantee order of evaluation, you could have two nested loops, resetting the array each time you start. Not sure which is fastest (your implementation resizes the array repeatedly while this alternative allocates/deallocates at least once too often), but at least, the alternative is less shocking.

Yoric replied:
Note: For once, I have no criticism about the comment itself :)

Gozala replied:
> Instead of this weird shift-based loop, if for some mysterious reason, you really want to guarantee order of evaluation

I tried to describe why this order is important, if you think guaranteed order needs to be justified in the comments to
appear less mysterious I'm open for suggestions. I was trying to avoid embedding novels in comments & can't think
of any concise explanation.

> you could have two nested loops, resetting the array each time you start. Not sure which is fastest (your implementation resizes the array repeatedly
> while this alternative allocates/deallocates at least once too often), but at least, the alternative is less shocking.

I'm afraid I don't agree this is common `dequeue` from `queue` until it's exhausted pattern, which to me reads way more intuitively than nested resetting arrays. I can't care about performance without bottlenecks & numbers. I'd like to wait @mossop on this, if he also finds this shocking than it's just me and I do whatever is less shocking.

Yoric replied:
Sorry, I didn't mean "shocking", I meant "surprising".

Mossop replied:
The main thing that strikes me as odd here is mixing the resolve/reject observers in the array. That might be confusing to new people coming to the code. I'd probably add an object `{resolve: onResolve, reject: onReject}` to the array instead and destructure from the shift. It might be slightly faster as you say, but how often are we actually going to have enough observers for that to matter?

The shift thing seems fine, though a single for loop is probably faster, but again unless we're going to see a lot of observers it is probably irrelevant.

Yoric replied:
So that things are clear: while I do not like the `shift`, my biggest problem is the double-`shift`.

Gozala replied:
Awesome, `{resolve: onResolve, reject: onReject}` solves it then.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+        while (observers.length)
+          result.then(observers.shift(), observers.shift());
+        // Once `observers` queue is exhausted we `null`-ify it, so that
+        // new handlers will be forwarded straight to the `result`.
```
"so that new handlers _are_ forwarded"

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,208 @@
+    reject: function reject(reason) {
+      // Note that if promise is resolved that does not necessary means that it
+      // is successfully fulfilled. Resolution value may be a promise in which
+      // case it's result propagates. In other words if promise `a` is resolved
```
Grammar: not "it's", "its".

Yoric replied:
Remember the rule: "it's" is the abbreviation of "it is". If you are unsure, replace all occurrences of "it's" by "it is". Then replace occurrences of "it is" that are obviously wrong by "its".

Gozala committed:
More suggested comment improvements.

Gozala commented:
## Update

At this point I think I have addressed all the comments with a few exceptions:
1. I have decided to defer cuddling else / catch style issues since that were Mozilla central and SDK style guides seem to have incompatible recommendations. We can come back to this once that's only thing left.
2. I insist that FIFO order of handler invocation must be guaranteed for several reasons:
   - Q that we claim to be a subset of has that, such a behavioral differences would be hard to spot and workaround.
   - Asynchronous computations expressed via promises will have FIFO order, this implies some expectations. 
     Invalidating those expectations when mixing with synchronous computations would make reasoning about code
     harder and switches from sync to async or wise versa problematic.
   - There are particular cases [described earlier](https://github.com/mozilla/mozilla-central/pull/4#discussion_r1004760)
     where no guarantees about order that makes proper solutions a lot more complicated. Also wrong assumptions about
     order won't be apparent to users and may become very tricky to spot.
   - Expectations about order is already set by an environment JS runs. Event listeners etc all have deterministic FIFO
     order new semantics will make sings only harder.
3. I tend to disagree that dequeue until exhausted pattern is confusing or harder to follow than nested resetting loops.
   There for I would like to get some other feedback to find out.
4. I got comments about observer pairs and it was not very clear if should have added comments or change. At the
   moment I have chose to documented why current approach is used instead of fork - join, although I'm open to change
   if that would be an improvement.

Other than that I think I have addressed everything else. Please highlight if I miss something.

Thanks!

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,212 @@
+        // forwarding are queued instead of being invoked immediately,
+        // guaranteeing FIFO order.
+        while (observers.length)
+          result.then(observers.shift(), observers.shift());
```
You seem to be changing this loop faster than I can comment on it, so I will just rollback to my previous criticisms:
- mixing `onResolve` and `onReject` in the same array is confusing, so please at the very least document this in the data structure – or even better, return to one of your previous versions that introduced an array or an object;
- these two uses of `shift` in the same expression are very confusing – you could afford the two lines to introduce named local variables, which would largely improve readability.

Gozala replied:
In fact I documented that, but anyway I'll just update to `{ resolve: onResolve, reject: onReject }` is was suggested.

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,212 @@
+
+  var deferred = {
+    promise: {
+      then: function then(onFulfill, onFailure) {
```
While I encourage you to not reuse the same variable/argument for two values with different definitions and types, using synonyms is almost as confusing.

So, I would go for `onResolve` => `decoratedOnResolve` => `propagateResolution`, for instance.

Gozala replied:
> using synonyms is almost as confusing.

In fact I changed names to highlight difference between promise **resolution** and **fulfillment**, I discovered when adding more comments that term _resolve_ was used for both making it confusing.

I do not like names `decorateOnResolve` and `propagateResolution` as they don't reflect the truth. Cause resulting function is
`decoratedOnResolve` only if `onResolve` is provided, otherwise it's an utility function that returns promise fulfilled / rejected with a given argument, which causes result to propagate. So choosing one of this name is actually misleading as it reader may be misguided causing him to ignore the other case. Accurate name would be `decoratedOnResolveOrPropagationFallbackIfNoOnResolve` or something like that, but of course that is too long and arguably less obvious than:

``` js
onFulfill = onFulfill ? attempt(onFulfill) : fulfillment;
```

Another reason why I prefer to reuse `onFullfill` here is because at the end of the day you just normalize provided handler that was passed, either by decorating it or by using fallback if missing. Also it clear for a reader that `onFulfill` is still a fulfill handle without reading  `attempt` or `fulfillment`. It is also clear that nothing other than `attempt` will ever call `onFulfill`. But if one would have assigned normalized handler to a fresh variable it would not be clear that original handlers are not used by anything except `attempt` without reading rest of the code (requiring at +1 register in readers mind).

Now this describes how I read and perceive code, if you both disagree I'm clearly doing it wrong and happy to change. Although I'd really love more accurate name, but unfortunately I can't think of anything short.

@Mossop what do you think ?

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,212 @@
+        // Store resolution `value` in a `result` as a promise, so that all
+        // the subsequent handlers can be simply forwarded to it. Since
+        // `result` will be a promise all the value / error propagation will
+        // be automatically taken care of.
```
I really think that you do not want to use word "automatically", but rather something like "uniformly".

Yoric commented on toolkit/addon-sdk/promise.js:
```diff
@@ -0,0 +1,212 @@
+ * fulfilled to a given `value`. Note the result is not a complete promise
+ * implementation, as its method `then` does not returns anything.
+ */
+function fulfillment(value) {
```
You just renamed `resolve` into `fulfillment`, is that it? This doesn't improve (or degrade) anything.

Why not something more descriptive, such as `asResolved`?

Gozala replied:
Because, I find it less descriptive and was hoping that native English @Mossop could help to figure out what would be most natural naming :)

Gozala committed:
Move module so that it will be in sync with SDK stuff once we start landing it.

Gozala commented:
# Update
- I have relocated module in order to make it compatible with SDK module layout, otherwise we'll run into conflicts when SDK modules start landing to mozilla-central.
-  I have made renames such that it makes most sense in my mind, although I'm not married to these names, it's just I don't like names that are [inaccurate](https://github.com/mozilla/mozilla-central/pull/4#discussion_r1013147)  or feel [less precise](https://github.com/mozilla/mozilla-central/pull/4#discussion_r1013223). I'd like to take advantage of the fact that @Mossop is native English, and let him choose / propose names that make most sense.
- I join resolve / reject observers into object it was suggested, which addresses multiple comment made in different parts of code.
- Added line breaks between comments to improve readability
- I still [don't buy that undeterministic order is better](https://github.com/mozilla/mozilla-central/pull/4#discussion_r1013025).
- I have created patch mozilla/addon-sdk#470 to make use this version in SDK and implement rest on top of it.
- Updated SDK promise test covering this library.

Mossop commented:
This pull request is getting long and arduous to read through. I'm going to skim over the naming for some of the functions and make some comments but once we're ok with those I suggest we land and take up remaining issues individually in other pull requests/bugs. Apologies if some of my comments have been addresser earlier but at goodness knows how many comments I'm bound to repeat you guys ;)

Mossop commented on toolkit/addon-sdk/promise/core.js:
```diff
@@ -0,0 +1,217 @@
+ */
+function fulfillment(value) {
+  return { then: function then(fulfill) { fulfill(value); } };
+}
```
This function returns a promise that is already resolved yes? Then I'd probably go with `resolvedPromise` or just `resolved` if you particularly want brevity (at the cost of a little less clarity of meaning)

Yoric replied:
This is not a real promise, so perhaps not `resolvedPromise`. A few months ago, ochameau suggested `resolvedFakePromise`.

Mossop replied:
Yeah I like `resolvedFakePromise`, makes it nice and clear what it is.

Mossop commented on toolkit/addon-sdk/promise/core.js:
```diff
@@ -0,0 +1,217 @@
+ */
+function rejection(reason) {
+  return { then: function then(fulfill, reject) { reject(reason); } };
+}
```
Same again, `rejectedPromise` or just `rejected`.

Mossop commented on toolkit/addon-sdk/promise/core.js:
```diff
@@ -0,0 +1,217 @@
+
+  var deferred = {
+    promise: {
+      then: function then(onFulfill, onError) {
```
I'd use onResolve and onReject here, that seems to be the terminology used elsewhere.

Gozala replied:
Well as a matter of fact there is a difference between promise resolution and fulfillment:

Promise **A** can be resolved with a promise **B**, which does not necessary means promise **A** will ever be fulfilled. As a matter of fact **A** will be rejected if **B** does. This handler is intentionally called `onFulfill` since it's only ever invoked if enclosed promise promise is fulfilled. Changing name to `onResolve` would be misleading since handler won't be called on promise resolution.

In addition this follows terminology in the spec more closely: http://wiki.commonjs.org/wiki/Promises/A

Mossop commented on toolkit/addon-sdk/promise/core.js:
```diff
@@ -0,0 +1,217 @@
+        // Create a pair of observers that invoke given handlers & propagate
+        // results to `deferred.promise`.
+        function resolve(value) { deferred.resolve(onFulfill(value)); }
+        function reject(reason) { deferred.resolve(onError(reason)); }
```
There are lots of repeated terms floating around, let's be clear and use `resolveDeferred` and `rejectDeferred` here.

Mossop commented on toolkit/addon-sdk/promise/core.js:
```diff
@@ -0,0 +1,217 @@
+  var deferred = defer();
+  deferred.reject(reason);
+  return deferred.promise;
+}
```
Actually maybe I'm missing something. How do these functions differ from the fulfillment and rejection functions above? Just more lightweight versions?

Yoric replied:
There is a considerable difference: `fulfillment` does not return an actual promise, but something that only looks like a promise (and cannot be chained). By opposition, `resolve` returns an actual promise.

I told you that that code was harder to understand than it looked :)

Gozala committed:
Address review comments.

Gozala commented:
I have addressed review comments with except changes from `fulfill` to `resolve` for reasons articulated above. Also between suggested `rejectedFakePromise` and `rejected` I choose later as the other one emphasises insignificant implementation detail. 

What's important about this function is that it takes value and boxes as rejection. The fact that it uses not standards compliant and even a promise is insignificant implementation detail, that is not worth distraction and in case of interest may be looked up. If you still think that `rejectedFakePromise` is better, I'm ok with changing it, just let me know.

fitzgen commented:
Hi, I am trying to use this library in a patch that I am writing for the debugger. I apologize if this is the wrong medium for this question, but I didn't get any answers in #jetpack.

I haven't been able to figure out where the library ends up after I build Nightly. I can't find it by browsing from "resource:///", nor can I find it via "find $OBJ_DIR -name core.js", and I can't seem to find any reference to it in any makefiles near where the file lives.

Thanks in advance.

Yoric commented:
For the moment, afaict, it does not end up anywhere. 
For my patches that depends on promise, I had to hack up something in the Makefile to grab core.js and embed it somewhere in resource://, until https://bugzilla.mozilla.org/show_bug.cgi?id=731779 lands.

ehsan commented:
Can this be closed now?

Yoric commented:
Yeah, let's close this issue.
